File: tests/wasm/test_keltner.js
/**
 * WASM binding tests for KELTNER indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? `file:///${wasmPath.replace(/\\/g, '/')}`
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('KELTNER accuracy - safe API', async () => {
    
    const { high, low, close } = testData;
    const source = close; 
    const { period, multiplier, ma_type } = EXPECTED_OUTPUTS.keltner.defaultParams;
    
    const result = wasm.keltner(high, low, close, source, period, multiplier, ma_type);
    
    assert.equal(result.rows, 3, 'Should have 3 output rows (upper, middle, lower)');
    assert.equal(result.cols, close.length, 'Output length should match input');
    
    const values = result.values;
    const len = close.length;
    
    
    const upper_band = values.slice(0, len);
    const middle_band = values.slice(len, 2 * len);
    const lower_band = values.slice(2 * len, 3 * len);
    
    
    const expectedUpper = [
        61619.504155205745,
        61503.56119134791,
        61387.47897150178,
        61286.61078267451,
        61206.25688331261
    ];
    const expectedMiddle = [
        59758.339871629956,
        59703.35512195091,
        59640.083205574636,
        59593.884805043715,
        59504.46720456336
    ];
    const expectedLower = [
        57897.17558805417,
        57903.14905255391,
        57892.68743964749,
        57901.158827412924,
        57802.67752581411
    ];
    
    assertArrayClose(upper_band.slice(-5), expectedUpper, 1e-1);
    assertArrayClose(middle_band.slice(-5), expectedMiddle, 1e-1);
    assertArrayClose(lower_band.slice(-5), expectedLower, 1e-1);
    
    
    const warmup = period - 1;
    for (let i = 0; i < warmup; i++) {
        assert(isNaN(upper_band[i]), `Upper band warmup value at ${i} should be NaN`);
        assert(isNaN(middle_band[i]), `Middle band warmup value at ${i} should be NaN`);
        assert(isNaN(lower_band[i]), `Lower band warmup value at ${i} should be NaN`);
    }
    
    
    try {
        await compareWithRust('keltner', upper_band, 'keltner', { period, multiplier, ma_type });
    } catch (e) {
        
    }
});

test('KELTNER default params', () => {
    
    const { high, low, close } = testData;
    const source = close;
    
    const result = wasm.keltner(high, low, close, source, 20, 2.0, "ema");
    
    assert.equal(result.rows, 3);
    assert.equal(result.cols, close.length);
    assert.equal(result.values.length, 3 * close.length);
});

test('KELTNER zero period', () => {
    
    const { high, low, close } = testData;
    const source = close;
    
    assert.throws(() => {
        wasm.keltner(high, low, close, source, 0, 2.0, "ema");
    }, /invalid period/i, 'Should throw error for period = 0');
});

test('KELTNER large period', () => {
    
    const { high, low, close } = testData;
    const source = close;
    
    assert.throws(() => {
        wasm.keltner(high, low, close, source, 999999, 2.0, "ema");
    }, /invalid period/i, 'Should throw error for period > data length');
});

test('KELTNER empty input', () => {
    
    const empty = new Float64Array(0);
    
    assert.throws(() => {
        wasm.keltner(empty, empty, empty, empty, 20, 2.0, "ema");
    }, /empty/i, 'Should throw error for empty data');
});

test('KELTNER very small dataset', () => {
    
    const smallData = new Float64Array([1.0, 2.0, 3.0]);
    
    assert.throws(() => {
        wasm.keltner(smallData, smallData, smallData, smallData, 20, 2.0, "ema");
    }, /invalid period|not enough valid data/i, 'Should throw error for insufficient data');
});

test('KELTNER all NaN values', () => {
    
    const len = 100;
    const nanData = new Float64Array(len).fill(NaN);
    
    assert.throws(() => {
        wasm.keltner(nanData, nanData, nanData, nanData, 20, 2.0, "ema");
    }, /all values are nan/i, 'Should throw error for all NaN values');
});

test('KELTNER invalid multiplier', () => {
    
    const data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0]);
    
    
    const negResult = wasm.keltner(data, data, data, data, 2, -2.0, "ema");
    assert.equal(negResult.values.length, 3 * data.length);
    
    
    const zeroResult = wasm.keltner(data, data, data, data, 2, 0.0, "ema");
    assert.equal(zeroResult.values.length, 3 * data.length);
    
    
    
    try {
        const nanResult = wasm.keltner(data, data, data, data, 2, NaN, "ema");
        
        const len = data.length;
        const upper = nanResult.values.slice(0, len);
        assert(upper.some(v => isNaN(v)), "Expected NaN in output for NaN multiplier");
    } catch (e) {
        
        assert(e.message.match(/invalid|NaN/), 'Should throw appropriate error for NaN multiplier');
    }
});

test('KELTNER different MA types', () => {
    
    const { high, low, close } = testData;
    const source = close.slice(0, 100); 
    const highSubset = high.slice(0, 100);
    const lowSubset = low.slice(0, 100);
    const closeSubset = close.slice(0, 100);
    
    const maTypes = ["ema", "sma", "wma", "rma"];
    
    for (const maType of maTypes) {
        try {
            const result = wasm.keltner(
                highSubset, lowSubset, closeSubset, source,
                20, 2.0, maType
            );
            assert.equal(result.cols, 100, `Failed for MA type: ${maType}`);
            assert.equal(result.rows, 3, `Failed for MA type: ${maType}`);
        } catch (e) {
            
            if (!e.message.toLowerCase().includes('unsupported')) {
                throw e;
            }
        }
    }
});

test('KELTNER NaN handling', () => {
    
    const { high, low, close } = testData;
    const source = close;
    const period = 20;
    
    const result = wasm.keltner(high, low, close, source, period, 2.0, "ema");
    
    const values = result.values;
    const len = close.length;
    
    
    const upper_band = values.slice(0, len);
    const middle_band = values.slice(len, 2 * len);
    const lower_band = values.slice(2 * len, 3 * len);
    
    
    if (len > 240) {
        for (let i = 240; i < len; i++) {
            assert(!isNaN(upper_band[i]), `Found unexpected NaN in upper band at index ${i}`);
            assert(!isNaN(middle_band[i]), `Found unexpected NaN in middle band at index ${i}`);
            assert(!isNaN(lower_band[i]), `Found unexpected NaN in lower band at index ${i}`);
        }
    }
    
    
    const warmup = period - 1;
    assertAllNaN(upper_band.slice(0, warmup), `Expected NaN in upper band warmup period (first ${warmup} values)`);
    assertAllNaN(middle_band.slice(0, warmup), `Expected NaN in middle band warmup period (first ${warmup} values)`);
    assertAllNaN(lower_band.slice(0, warmup), `Expected NaN in lower band warmup period (first ${warmup} values)`);
});

test('KELTNER batch single parameter set', () => {
    
    const { high, low, close } = testData;
    const source = close;
    const period = 20;
    const multiplier = 2.0;
    
    
    const singleResult = wasm.keltner(high, low, close, source, period, multiplier, "ema");
    
    
    const config = {
        period_range: [period, period, 0],
        multiplier_range: [multiplier, multiplier, 0],
        ma_type: "ema"
    };
    
    const batchResult = wasm.keltner_batch(high, low, close, source, config);
    
    
    assert.equal(batchResult.rows, 1);
    assert.equal(batchResult.cols, close.length);
    assert.equal(batchResult.combos.length, 1);
    
    
    assert.equal(batchResult.combos[0].period, period);
    assert.equal(batchResult.combos[0].multiplier, multiplier);
    
    
    const len = close.length;
    const singleUpper = singleResult.values.slice(0, len);
    const singleMiddle = singleResult.values.slice(len, 2 * len);
    const singleLower = singleResult.values.slice(2 * len, 3 * len);
    
    const batchUpper = batchResult.upper.slice(0, len);
    const batchMiddle = batchResult.middle.slice(0, len);
    const batchLower = batchResult.lower.slice(0, len);
    
    assertArrayClose(batchUpper, singleUpper, 1e-10, "Batch upper doesn't match single");
    assertArrayClose(batchMiddle, singleMiddle, 1e-10, "Batch middle doesn't match single");
    assertArrayClose(batchLower, singleLower, 1e-10, "Batch lower doesn't match single");
});

test('KELTNER batch multiple parameters', () => {
    
    const { high, low, close } = testData;
    const source = close.slice(0, 100); 
    const highSubset = high.slice(0, 100);
    const lowSubset = low.slice(0, 100);
    const closeSubset = close.slice(0, 100);
    
    const config = {
        period_range: [10, 30, 10],  
        multiplier_range: [1.0, 3.0, 1.0],  
        ma_type: "ema"
    };
    
    const result = wasm.keltner_batch(highSubset, lowSubset, closeSubset, source, config);
    
    
    assert.equal(result.rows, 9, 'Should have 9 parameter combinations');
    assert.equal(result.cols, 100, 'Output columns should match input length');
    assert.equal(result.combos.length, 9, 'Should have 9 parameter combinations');
    
    
    assert.equal(result.combos[0].period, 10);
    assert.equal(result.combos[0].multiplier, 1.0);
    
    
    assert.equal(result.combos[8].period, 30);
    assert.equal(result.combos[8].multiplier, 3.0);
    
    
    assert.equal(result.upper.length, 9 * 100);
    assert.equal(result.middle.length, 9 * 100);
    assert.equal(result.lower.length, 9 * 100);
    
    
    const firstRowUpper = result.upper.slice(0, 100);
    const firstRowMiddle = result.middle.slice(0, 100);
    const firstRowLower = result.lower.slice(0, 100);
    
    const singleResult = wasm.keltner(
        highSubset, lowSubset, closeSubset, source,
        10, 1.0, "ema"
    );
    
    const singleUpper = singleResult.values.slice(0, 100);
    const singleMiddle = singleResult.values.slice(100, 200);
    const singleLower = singleResult.values.slice(200, 300);
    
    assertArrayClose(firstRowUpper, singleUpper, 1e-10, "First batch row upper doesn't match single");
    assertArrayClose(firstRowMiddle, singleMiddle, 1e-10, "First batch row middle doesn't match single");
    assertArrayClose(firstRowLower, singleLower, 1e-10, "First batch row lower doesn't match single");
});

test('KELTNER batch metadata', () => {
    
    const data = new Float64Array(50);
    data.fill(100);
    
    const config = {
        period_range: [10, 20, 5],  
        multiplier_range: [1.0, 2.0, 0.5],  
        ma_type: "ema"
    };
    
    const result = wasm.keltner_batch(data, data, data, data, config);
    
    
    assert.equal(result.combos.length, 9);
    
    
    const expectedCombos = [
        { period: 10, multiplier: 1.0 },
        { period: 10, multiplier: 1.5 },
        { period: 10, multiplier: 2.0 },
        { period: 15, multiplier: 1.0 },
        { period: 15, multiplier: 1.5 },
        { period: 15, multiplier: 2.0 },
        { period: 20, multiplier: 1.0 },
        { period: 20, multiplier: 1.5 },
        { period: 20, multiplier: 2.0 }
    ];
    
    for (let i = 0; i < 9; i++) {
        assert.equal(result.combos[i].period, expectedCombos[i].period, 
                     `Period mismatch at combo ${i}`);
        assertClose(result.combos[i].multiplier, expectedCombos[i].multiplier, 1e-10,
                    `Multiplier mismatch at combo ${i}`);
    }
});

test('KELTNER batch full parameter sweep', () => {
    
    const data = new Float64Array(50);
    data.fill(100);
    
    const config = {
        period_range: [10, 14, 2],      
        multiplier_range: [1.0, 1.5, 0.5],  
        ma_type: "ema"
    };
    
    const batchResult = wasm.keltner_batch(data, data, data, data, config);
    
    
    assert.equal(batchResult.combos.length, 6);
    assert.equal(batchResult.rows, 6);
    assert.equal(batchResult.cols, 50);
    assert.equal(batchResult.upper.length, 6 * 50);
    assert.equal(batchResult.middle.length, 6 * 50);
    assert.equal(batchResult.lower.length, 6 * 50);
});

test.after(() => {
    console.log('KELTNER WASM tests completed');
});

File: tests/python/test_keltner.py
"""
Python binding tests for KELTNER indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestKeltner:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_keltner_accuracy(self, test_data):
        """Test KELTNER matches expected values from Rust tests - mirrors check_keltner_accuracy"""
        expected = EXPECTED_OUTPUTS['keltner']


        upper_band, middle_band, lower_band = ta_indicators.keltner(
            test_data['high'],
            test_data['low'],
            test_data['close'],
            test_data['close'],
            expected['default_params']['period'],
            expected['default_params']['multiplier'],
            expected['default_params']['ma_type']
        )


        assert len(upper_band) == len(test_data['close'])
        assert len(middle_band) == len(test_data['close'])
        assert len(lower_band) == len(test_data['close'])


        assert_close(
            upper_band[-5:],
            expected['last_5_upper'],
            rtol=1e-8,
            msg="Upper band mismatch"
        )

        assert_close(
            middle_band[-5:],
            expected['last_5_middle'],
            rtol=1e-8,
            msg="Middle band mismatch"
        )

        assert_close(
            lower_band[-5:],
            expected['last_5_lower'],
            rtol=1e-8,
            msg="Lower band mismatch"
        )


        try:
            compare_with_rust('keltner', upper_band, 'keltner', expected['default_params'])
        except:
            pass

    def test_keltner_default_params(self, test_data):
        """Test KELTNER with default parameters - mirrors check_keltner_default_params"""

        upper, middle, lower = ta_indicators.keltner(
            test_data['high'],
            test_data['low'],
            test_data['close'],
            test_data['close'],
            20, 2.0, "ema"
        )

        assert len(upper) == len(test_data['close'])
        assert len(middle) == len(test_data['close'])
        assert len(lower) == len(test_data['close'])

    def test_keltner_zero_period(self, test_data):
        """Test KELTNER fails with zero period - mirrors check_keltner_zero_period"""
        with pytest.raises(ValueError, match="invalid period"):
            ta_indicators.keltner(
                test_data['high'],
                test_data['low'],
                test_data['close'],
                test_data['close'],
                0, 2.0, "ema"
            )

    def test_keltner_large_period(self, test_data):
        """Test KELTNER fails when period exceeds data length - mirrors check_keltner_large_period"""
        with pytest.raises(ValueError, match="invalid period"):
            ta_indicators.keltner(
                test_data['high'],
                test_data['low'],
                test_data['close'],
                test_data['close'],
                999999, 2.0, "ema"
            )

    def test_keltner_empty_input(self):
        """Test KELTNER fails with empty input - similar to ALMA's check_alma_empty_input"""
        empty = np.array([])

        with pytest.raises(ValueError, match="empty data"):
            ta_indicators.keltner(empty, empty, empty, empty, 20, 2.0, "ema")

    def test_keltner_very_small_dataset(self):
        """Test KELTNER with insufficient data - similar to ALMA's check_alma_very_small_dataset"""
        small_data = np.array([1.0, 2.0, 3.0])

        with pytest.raises(ValueError, match="invalid period|not enough valid data"):
            ta_indicators.keltner(
                small_data, small_data, small_data, small_data,
                20, 2.0, "ema"
            )

    def test_keltner_all_nan_input(self):
        """Test KELTNER with all NaN values"""
        all_nan = np.full(100, np.nan)

        with pytest.raises(ValueError, match="[Aa]ll values are"):
            ta_indicators.keltner(
                all_nan, all_nan, all_nan, all_nan,
                20, 2.0, "ema"
            )

    def test_keltner_invalid_multiplier(self):
        """Test KELTNER with invalid multiplier values"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])


        upper, middle, lower = ta_indicators.keltner(
            data, data, data, data,
            2, -2.0, "ema"
        )
        assert len(upper) == len(data)


        upper, middle, lower = ta_indicators.keltner(
            data, data, data, data,
            2, 0.0, "ema"
        )
        assert len(upper) == len(data)



        try:
            upper, middle, lower = ta_indicators.keltner(
                data, data, data, data,
                2, float('nan'), "ema"
            )

            assert np.all(np.isnan(upper[2:])), "Expected NaN output for NaN multiplier"
        except ValueError:

            pass

    def test_keltner_different_ma_types(self, test_data):
        """Test KELTNER with different MA types"""
        ma_types = ["ema", "sma", "wma", "rma"]

        for ma_type in ma_types:
            try:
                upper, middle, lower = ta_indicators.keltner(
                    test_data['high'][:100],
                    test_data['low'][:100],
                    test_data['close'][:100],
                    test_data['close'][:100],
                    20, 2.0, ma_type
                )
                assert len(upper) == 100, f"Failed for MA type: {ma_type}"
                assert len(middle) == 100, f"Failed for MA type: {ma_type}"
                assert len(lower) == 100, f"Failed for MA type: {ma_type}"
            except ValueError as e:

                if "unsupported" not in str(e).lower():
                    raise

    def test_keltner_nan_handling(self, test_data):
        """Test KELTNER handles NaN values correctly - mirrors check_keltner_nan_handling"""
        period = 20
        upper, middle, lower = ta_indicators.keltner(
            test_data['high'],
            test_data['low'],
            test_data['close'],
            test_data['close'],
            period, 2.0, "ema"
        )

        assert len(upper) == len(test_data['close'])
        assert len(middle) == len(test_data['close'])
        assert len(lower) == len(test_data['close'])


        if len(upper) > 240:
            assert not np.any(np.isnan(upper[240:])), "Found unexpected NaN in upper band after warmup"
            assert not np.any(np.isnan(middle[240:])), "Found unexpected NaN in middle band after warmup"
            assert not np.any(np.isnan(lower[240:])), "Found unexpected NaN in lower band after warmup"


        warmup = period - 1
        assert np.all(np.isnan(upper[:warmup])), f"Expected NaN in upper band warmup period (first {warmup} values)"
        assert np.all(np.isnan(middle[:warmup])), f"Expected NaN in middle band warmup period (first {warmup} values)"
        assert np.all(np.isnan(lower[:warmup])), f"Expected NaN in lower band warmup period (first {warmup} values)"

    def test_keltner_streaming(self, test_data):
        """Test streaming functionality - mirrors check_keltner_streaming"""
        expected = EXPECTED_OUTPUTS['keltner']


        batch_upper, batch_middle, batch_lower = ta_indicators.keltner(
            test_data['high'],
            test_data['low'],
            test_data['close'],
            test_data['close'],
            expected['default_params']['period'],
            expected['default_params']['multiplier'],
            expected['default_params']['ma_type']
        )


        stream = ta_indicators.KeltnerStream(
            expected['default_params']['period'],
            expected['default_params']['multiplier'],
            expected['default_params']['ma_type']
        )


        stream_upper = []
        stream_middle = []
        stream_lower = []

        for i in range(len(test_data['close'])):
            result = stream.update(
                test_data['high'][i],
                test_data['low'][i],
                test_data['close'][i],
                test_data['close'][i]
            )
            if result is not None:
                stream_upper.append(result[0])
                stream_middle.append(result[1])
                stream_lower.append(result[2])
            else:
                stream_upper.append(np.nan)
                stream_middle.append(np.nan)
                stream_lower.append(np.nan)

        stream_upper = np.array(stream_upper)
        stream_middle = np.array(stream_middle)
        stream_lower = np.array(stream_lower)


        for i, (b_u, s_u, b_m, s_m, b_l, s_l) in enumerate(zip(
            batch_upper, stream_upper, batch_middle, stream_middle, batch_lower, stream_lower
        )):
            if np.isnan(b_u) and np.isnan(s_u):
                continue
            assert_close(b_u, s_u, rtol=1e-8, atol=1e-8,
                        msg=f"Upper band streaming mismatch at index {i}")
            assert_close(b_m, s_m, rtol=1e-8, atol=1e-8,
                        msg=f"Middle band streaming mismatch at index {i}")
            assert_close(b_l, s_l, rtol=1e-8, atol=1e-8,
                        msg=f"Lower band streaming mismatch at index {i}")

    def test_keltner_batch_default_row(self, test_data):
        """Test batch with default parameters matches single calculation - mirrors check_batch_default_row"""
        expected = EXPECTED_OUTPUTS['keltner']


        single_upper, single_middle, single_lower = ta_indicators.keltner(
            test_data['high'],
            test_data['low'],
            test_data['close'],
            test_data['close'],
            expected['default_params']['period'],
            expected['default_params']['multiplier'],
            expected['default_params']['ma_type']
        )


        result = ta_indicators.keltner_batch(
            test_data['high'],
            test_data['low'],
            test_data['close'],
            test_data['close'],
            (20, 20, 0),
            (2.0, 2.0, 0.0)
        )

        assert 'upper' in result
        assert 'middle' in result
        assert 'lower' in result
        assert 'periods' in result
        assert 'multipliers' in result


        assert result['upper'].shape[0] == 1
        assert result['middle'].shape[0] == 1
        assert result['lower'].shape[0] == 1
        assert result['upper'].shape[1] == len(test_data['close'])


        batch_upper_row = result['upper'][0]
        batch_middle_row = result['middle'][0]
        batch_lower_row = result['lower'][0]


        assert_close(batch_upper_row, single_upper, rtol=1e-8,
                    msg="Batch upper band doesn't match single calculation")
        assert_close(batch_middle_row, single_middle, rtol=1e-8,
                    msg="Batch middle band doesn't match single calculation")
        assert_close(batch_lower_row, single_lower, rtol=1e-8,
                    msg="Batch lower band doesn't match single calculation")

    def test_keltner_batch_multiple_params(self, test_data):
        """Test batch functionality with multiple parameter combinations"""

        high = test_data['high'][:100]
        low = test_data['low'][:100]
        close = test_data['close'][:100]

        result = ta_indicators.keltner_batch(
            high, low, close, close,
            (10, 30, 10),
            (1.0, 3.0, 1.0)
        )


        expected_rows = 3 * 3
        assert result['upper'].shape == (expected_rows, 100)
        assert result['middle'].shape == (expected_rows, 100)
        assert result['lower'].shape == (expected_rows, 100)


        assert len(result['periods']) == expected_rows
        assert len(result['multipliers']) == expected_rows


        assert result['periods'][0] == 10
        assert result['multipliers'][0] == 1.0
        assert result['periods'][-1] == 30
        assert result['multipliers'][-1] == 3.0



        single_upper, single_middle, single_lower = ta_indicators.keltner(
            high, low, close, close, 10, 1.0, "ema"
        )

        assert_close(result['upper'][0], single_upper, rtol=1e-8,
                    msg="First batch row doesn't match single calculation")


if __name__ == '__main__':
    pytest.main([__file__, '-v'])